File: year2020/day04/python/solution.py
import re

def main():
    with open('../puzzle.txt') as file:
        input = [passport.split() for passport in file.read().split('\n\n')]
    
    print(part_1(input))
    print(part_2(input))


def part_1(input):
    required = ['byr', 'iyr', 'eyr', 'hgt', 'hcl', 'ecl', 'pid']
    return len([passport for passport in input if all(field in {seq.split(":")[0] for seq in passport} for field in required)])


def part_2(input):
    return len([passport for passport in input if validate(dict([entry.split(':') for entry in passport]))])


def validate(passport_dict):
    fields = {
        'byr': (1920, 2002),
        'iyr': (2010, 2020),
        'eyr': (2020, 2030),
        'hcl': r'^#[0-9a-f]{6}$',
        'ecl': ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'],
        'pid': r'^\d{9}$'
    }

    for field, rule in fields.items():
        if field not in passport_dict:
            return False
        
        if isinstance(rule, tuple):
            if not (rule[0] <= int(passport_dict[field]) <= rule[1]):
                return False
            
        elif isinstance(rule, str):
            if not re.match(rule, passport_dict[field]):
                return False
            
        elif isinstance(rule, list):
            if passport_dict[field] not in rule:
                return False
    
    if 'hgt' in passport_dict:
        hgt_value = passport_dict['hgt'][:-2]
        hgt_unit = passport_dict['hgt'][-2:]
        if hgt_unit == "cm" and 150 <= int(hgt_value) <= 193:
            return True
        
        elif hgt_unit == "in" and 59 <= int(hgt_value) <= 76:
            return True
    
    return False


if __name__ == '__main__':
    main()

File: year2020/day04/java/Solution.java
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;
import java.util.stream.Collectors;

public class Solution {
  private static String[][] PASSPORTS;
  private static List<String> REQUIRED = Arrays.asList("byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid");

  public static void main(String[] args) {
    try {
      setup(Files.readString(Paths.get("../puzzle.txt")));
      System.out.println(part1());
      System.out.println(part2());
    } catch(Exception e) {
      e.printStackTrace();
    }
  }

	public static int part1() {
    return (int) Arrays.stream(PASSPORTS)
        .filter(passport -> REQUIRED.stream()
        .allMatch(field -> Arrays.stream(passport)
        .map(seq -> seq.split(":")[0])
        .collect(Collectors.toSet())
        .contains(field)))
        .count();
  }

  public static int part2() {
    return (int) Arrays.stream(PASSPORTS).filter(passport -> validate(parsePassportFields(passport))).count();
  }

  public static void setup(String input) {
    PASSPORTS = Arrays.stream(input.split("\n\n")).map(passport -> passport.split("\\s+")).toArray(String[][]::new);
  }

  private static Map<String, String> parsePassportFields(String[] passport) {
    Map<String, String> dict = new HashMap<>();
    for (String field : passport) {
      String[] splitField = field.split(":");
      dict.put(splitField[0], splitField.length > 1 ? splitField[1] : null);
    }
    return dict;
  }

  private static boolean validate(Map<String, String> dict) {
    try {
      if (!(dict.containsKey("byr") && 1920 <= Integer.parseInt(dict.get("byr")) && Integer.parseInt(dict.get("byr")) <= 2002))
        return false;
      if (!(dict.containsKey("iyr") && 2010 <= Integer.parseInt(dict.get("iyr")) && Integer.parseInt(dict.get("iyr")) <= 2020))
        return false;
      if (!(dict.containsKey("eyr") && 2020 <= Integer.parseInt(dict.get("eyr")) && Integer.parseInt(dict.get("eyr")) <= 2030))
        return false;
      if (!(dict.containsKey("hcl") && Pattern.matches("^#[0-9a-f]{6}$", dict.get("hcl"))))
        return false;
      if (!(dict.containsKey("ecl") && Arrays.asList("amb", "blu", "brn", "gry", "grn", "hzl", "oth").contains(dict.get("ecl"))))
        return false;
      if (!(dict.containsKey("pid") && Pattern.matches("^\\d{9}$", dict.get("pid"))))
        return false;

      if (dict.containsKey("hgt")) {
        String hgt = dict.get("hgt");
        int height = Integer.parseInt(hgt.replaceAll("[^0-9]", ""));
        if (hgt.endsWith("cm") && (150 <= height && height <= 193))
          return true;
        if (hgt.endsWith("in") && (59 <= height && height <= 76))
          return true;
      }

    } catch (NumberFormatException e) {
      return false;
    }

    return false;
  }
}